### DISCUSSION TITLE
How to globally change the uid lib?

### DISCUSSION THREAD

**Discussion by MartinX3** [CodeDredd/pinia-orm#1819] [Q&A]
I just want to extends from `model` without defining the `@Uid` field in the entities since it will be auto generated.

My used framework quarkus has already a buildin uid feature so I wouldn't need to install another package.

See
https://quasar.dev/quasar-utils/other-utils#uid-generate-uid
https://github.com/quasarframework/quasar/blob/dev/ui/src/utils/uid/uid.js

**Comment by CodeDredd** [comment]
Well you can't overwrite the `@Uid` decorator but you can create your own one. Just create your custom cast and decorator.
Look here https://pinia-orm.codedredd.de/guide/model/casts#defining-custom-casts
Or look there https://github.com/CodeDredd/pinia-orm/tree/main/packages/pinia-orm/src/packages/uuid

**Reply by MartinX3** [reply]
But would that not need to define a `@Uid` variable in every single class extending from `Model`?
I was hoping I could change the automatically created id field using a custom uid generator.

**Reply by CodeDredd** [reply]
Yes you would need to define your own decorator.

Where are many ways to implement what you need. It depends on how the rest of your data architecture is.
You can define a `BaseModel` where you can define this one uid field which you can populate now with either mutators or casts or direct value setting...

**Reply by MartinX3** [reply]
That's correct by using a abstract base class.  
I just thought i could place a class with a decorator somewhere to overwrite the default @Uid implementation like in Spring Boot (JVM) so I don't need that extra base class and just use the provided model class.  
Or using an option to provide an absolute path of my own implementation class to the framework to use it instead of the default implementation. :D

**Reply by CodeDredd** [reply]
I know what you mean. In PHP it's also easy to overwrite a namespace. But in JS it's not that "native". I think there libriaries in JS which can do that. But that feels all not right.
Maybe i implement some mapping for the uid in the config so you can overwrite it that way.

**Reply by MartinX3** [reply]
Thank you.

And another question about UID i have.
I see that I need to install `nanoid`.
What about using the buildin uuid generator by default?
`crypto.randomUUID()` for v4 UUIDs exists since years.
See https://developer.mozilla.org/en-US/docs/web/api/crypto/randomuuid#browser_compatibility

**Reply by CodeDredd** [reply]
No you don't need to install nanoid. You can choose what you like. 
`crypto` is not always avaible. So thats the reason not just to use it.
The standard uid uses the same logic as [nanoid/non-secure](https://github.com/ai/nanoid?tab=readme-ov-file#non-secure).
But if you like you can also use uuid instead. https://pinia-orm.codedredd.de/guide/model/getting-started#uid-type

**Reply by MartinX3** [reply]
Oh, I thought it was available everywhere because of the big browser list at this page.

Yeah, I just thought I could avoid to install another package. :D
